Visualize cancelled state

boyvanamstel opened the pull request:
Defaults to desaturating the progress view when it's cancelled.

Still need to implement the desaturated color animating the color change.

Fixes https://github.com/sindresorhus/CircularProgress/issues/6

boyvanamstel committed:
Add option to change color when cancelled

sindresorhus commented on Sources/CircularProgress/util.swift:
```diff
@@ -30,6 +29,10 @@ extension NSColor {
+
+	var desaturated: NSColor {
+		return NSColor.black // TODO: Desaturate color
+	}
```
In my private framework, I have this implemented as:

```swift
extension NSColor {
	typealias HSBAColor = (hue: Double, saturation: Double, brightness: Double, alpha: Double)
	var hsba: HSBAColor {
		var hue: CGFloat = 0
		var saturation: CGFloat = 0
		var brightness: CGFloat = 0
		var alpha: CGFloat = 0
		let color = usingColorSpace(.deviceRGB) ?? self
		color.getHue(&hue, saturation: &saturation, brightness: &brightness, alpha: &alpha)
		return HSBAColor(Double(hue), Double(saturation), Double(brightness), Double(alpha))
	}
}

extension NSColor {
	private func colorWithSaturation(ratio: Double) -> NSColor {
		let color = hsba
		return NSColor(
			hue: CGFloat(color.hue),
			saturation: CGFloat(color.saturation * ratio),
			brightness: CGFloat(color.brightness),
			alpha: CGFloat(color.alpha)
		)
	}

	func desaturating(by ratio: Double) -> NSColor {
		return colorWithSaturation(ratio: 1 - ratio)
	}
}
```

Feel free to borrow (or improve) that implementation.

boyvanamstel replied:
I have something similar, but I'll use yours. Thanks. 👍

sindresorhus replied:
Please let me know if your implementation does something better, in your opinion. I'm always looking to improve my utility methods.

boyvanamstel committed:
Implement desaturation and fallback

boyvanamstel committed:
Replace desaturation extension

boyvanamstel committed:
Add animation and custom color generation handler

boyvanamstel commented:
I've added two options for `.cancelledStateVisualizationMethod`: `.desaturate` and `.disable`. Besides that I've added a `.cancelledStateColorHandler` that provides the current color and allows the user to return how it should look when it's cancelled.

``` swift
progressCircularProgress.cancelledStateColorHandler = { color in
	return color.withAlphaComponent(0.2)
}
```

I'll add it all to the documentation when we've agreed on the implementation.

boyvanamstel commented:
![desaturate](https://user-images.githubusercontent.com/225410/53086073-a8d23300-3504-11e9-8158-6de041830ae6.gif)

I'm currently desaturating by 70% and it's a bit hit and miss. The `.disable` method is a bit more standard.

![disable](https://user-images.githubusercontent.com/225410/53086296-1d0cd680-3505-11e9-9aa1-63bb08f89d28.gif)

The callback should give maximum flexibility. I could make it a property as well. The benefit of having a callback is that it works properly when changing the color of the progress view (like we do in the example). Otherwise a property would be sufficient.

sindresorhus commented:
Hmm, you're right. Maybe desaturation was not the best idea. Since the point is to show that it's cancelled, and the usual way to do that is to give it less attention, but by desaturating it, it becomes lighter, which gives it more attention. Maybe we should just turn it into a boolean instead for disable or not?

sindresorhus commented:
> The callback should give maximum flexibility.

👍

sindresorhus commented on Sources/CircularProgress/CustomButton.swift:
```diff
@@ -71,14 +79,19 @@ extension CALayer {
 		animation.fillMode = .forwards
+		animation.delegate = animationDelegate
 		animation.isRemovedOnCompletion = false
 		add(animation, forKey: keyPath)
```
You could add something like this extension to simplify it further:

```swift
extension CALayer {
	func add(_ animation: CAAnimation, forKey key: String?, completion: ((Bool) -> Void)?) {
		let animationDelegate = AnimationDelegate()
		animationDelegate.didStopHandler = completion
		animation.delegate = animationDelegate
		add(animation, forKey: key)
	}
}
```

boyvanamstel replied:
Great idea. 👍

sindresorhus commented on Sources/CircularProgress/CustomButton.swift:
```diff
@@ -63,6 +63,14 @@ final class TrackingArea {
+final class AnimationDelegate: NSObject, CAAnimationDelegate {
+	var didStopHandler: ((Bool) -> Void)?
+
+	func animationDidStop(_ anim: CAAnimation, finished flag: Bool) {
```
```suggestion
	func animationDidStop(_ animation: CAAnimation, finished flag: Bool) {
```

boyvanamstel replied:
Ugh, fixing.

boyvanamstel committed:
Upgrade adding animations

boyvanamstel commented:
> Hmm, you're right. Maybe desaturation was not the best idea. Since the point is to show that it's cancelled, and the usual way to do that is to give it less attention, but by desaturating it, it becomes lighter, which gives it more attention. Maybe we should just turn it into a boolean instead for disable or not?

You mean remove the desaturation and only go with the disabled color? I'm for that.

boyvanamstel commented:
![desaturate-v2](https://user-images.githubusercontent.com/225410/53171810-0d17f400-35e4-11e9-8c97-9474316f1b4e.gif)

Or we manually set the `brightness` to something more suitable: `color.desaturating(by: 0.7, brightness: 0.5)`.

boyvanamstel committed:
Expose desaturation properties

boyvanamstel commented:
I now expose the saturation settings through `.desaturate(properties: (ratio: 0.7, brightness: 0.5))`.

---

The current implementation works around the warning below. Is this deliberate?

<img width="1768" alt="screenshot 2019-02-21 at 14 31 40" src="https://user-images.githubusercontent.com/225410/53172409-8e23bb00-35e5-11e9-850e-490feb8184dd.png">

sindresorhus commented:
> You mean remove the desaturation and only go with the disabled color? I'm for that.

Yes

boyvanamstel commented:
> Yes

See the new options. I think that's an even better solution than getting rid of the desaturation.

sindresorhus commented:
> Is this deliberate?

You need to move the `let` to the outside:

```swift
case let .desaturate(ratio, brightness):
```

boyvanamstel committed:
Tweak desaturation case syntax

boyvanamstel committed:
Update readme

sindresorhus commented:
Can you run the GIF through ImageOptim.app?

sindresorhus commented on Sources/CircularProgress/CustomButton.swift:
```diff
@@ -77,8 +85,16 @@ extension CALayer {
+		}
+	}
+
+	func add(_ animation: CAAnimation, forKey key: String?, completion: ((Bool) -> Void)?) {
```
Actually, no point in making the closure optional as then it overlaps with the built-in method.

```suggestion
	func add(_ animation: CAAnimation, forKey key: String?, completion: ((Bool) -> Void)) {
```

boyvanamstel replied:
Good catch.

sindresorhus commented:
I have been experimenting with different combinations and this is my favorite:

```swift
alphaValue = 0.6
color = color.desaturating(by: 0.4, brightness: 0.8)
```

By reducing the opacity too, it becomes clear enough that it's in a cancelled/disabled state.

So my thinking is that we make the enum:
- `.none`
- `.disabled` (which would be the above values)
- `.custom(opacity: Double, saturation: Double, brightness: Double)` (no defaults)

*Note: `disabled`, not `disable`.*

With `disabled` being the default. That way we have a good default, but also have it customizable.

Thoughts?

sindresorhus commented:
Hmm, I'm starting to think that we are overthinking this. How about just a `. visualizeCancelledState: Bool` property that uses my suggested style [1] by default? I can't imagine anyone would bother to customize it, and if they do, they can set the custom color in `.cancelledStateColorHandler`.

[1]: `alphaValue = 0.6; color = color.desaturating(by: 0.4, brightness: 0.8)`

boyvanamstel commented:
> Hmm, I'm starting to think that we are overthinking this. How about just a `. visualizeCancelledState: Bool` property that uses my suggested style [1] by default? I can't imagine anyone would bother to customize it, and if they do, they can set the custom color in `.cancelledStateColorHandler`.
> 
> [1]: `alphaValue = 0.6; color = color.desaturating(by: 0.4, brightness: 0.8)`

I agree. The `.cancelledStateColorHandler` is enough of a fallback for any other options people might want. Implementing now.

boyvanamstel committed:
Simplify cancelled state visualization

boyvanamstel committed:
Make animation callback non-optional

boyvanamstel commented:
I've finished making the changes. One thing to note is that the default visualization works best in Dark Mode. I use Light Mode normally and the lighter colors completely disappear when the opacity is changed.

This is obviously a side-effect of the Example application's use of colors, but to offer some flexibility I skip the opacity change when the `.cancelledStateColorHandler` is set. That way users can optionally change the opacity themselves in combination with their color change.

I've mentioned this behavior in the manual.

sindresorhus committed:
Update AppDelegate.swift

sindresorhus committed:
Update CircularProgress.swift

sindresorhus commented:
> I've finished making the changes. One thing to note is that the default visualization works best in Dark Mode. I use Light Mode normally and the lighter colors completely disappear when the opacity is changed.

I noticed that too. I've adjusted the opacity from 0.6 to 0.7.

sindresorhus commented:
> This is obviously a side-effect of the Example application's use of colors, but to offer some flexibility I skip the opacity change when the `.cancelledStateColorHandler` is set. That way users can optionally change the opacity themselves in combination with their color change.

👍

sindresorhus commented:
This is great. Thanks for implementing it 👌

boyvanamstel commented:
You're welcome!
